src/Channel.cpp

```diff
@@ -84,7 +87,85 @@ void Channel::remove(Channel *c) {
 	QWriteLocker lock(&c_qrwlChannels);
 	c_qhChannels.remove(c->iId);
 }
-#endif
+
+void Channel::setFilterMode(Channel::FilterMode filterMode) {
+	m_filterMode = filterMode;
+	ServerHandlerPtr sh = Global::get().sh;
+	if (sh) {
+		Global::get().db->setChannelFilterMode(sh->qbaDigest, iId, m_filterMode);
+	}
+}
+
+bool Channel::isFiltered() const {
+	if (!Global::get().s.bFilterActive) {
+		// Channel filtering is disabled
+		return false;
+	}
+
+	Channel *userChannel = nullptr;
+
+	if (Global::get().uiSession != 0) {
+		const ClientUser *user = ClientUser::get(Global::get().uiSession);
+		if (user) {
+			userChannel = user->cChannel;
+		}
+	}
+
+	bool hasUser = false;
+	bool hasHide = false;
+	bool hasPin  = false;
+
+	// Iterate tree down
+	std::queue< const Channel * > channelSearch;
+	channelSearch.push(this);
+
+	while (!channelSearch.empty()) {
+		const Channel *c = channelSearch.front();
+		channelSearch.pop();
+
+		// Never hide channel, if user resides in this channel or a subchannel
+		if (userChannel != nullptr && c == userChannel) {
+			return false;
+		}
+
+		if (!c->qlUsers.isEmpty()) {
+			hasUser = true;
+		}
+
+		if (c->m_filterMode == Channel::FILTER_MODE_PIN) {
+			hasPin = true;
+		}
```

Comment:
I would think that finding either of this could terminate the loop, couldn't it?

Reply:
Yes you are right I will change that. But remember: if the filter logic ever changes these breaks will be hard to spot and could cause very interesting bugs.

Reply:
See, I just switched ``hasPin`` and ``hasHide`` to enable to hide a parent channel, which then hides all the child channels except pinned one and the break in ``hasUser`` broke that change.

Reply:
> set the flags no matter what 

I think I don't quite understand what you mean by that :thinking:

Reply:
Basically, do not break there such that the flags ``hasUser``, ``hasPin`` and ``hasHide`` are correctly set after the loops. Otherwise, if someone was to ever change the logic they could not rely on these flags and would probably be very confused.

Reply:
Ah, I see. Yeah that sounds reasonable and this code is not performance critical and thus, it should be fine to perform a few extra loop iterations.